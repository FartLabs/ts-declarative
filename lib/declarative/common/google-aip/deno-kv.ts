import {
  standardCreate,
  standardDelete,
  standardGet,
  standardList,
  standardUpdate,
} from "./methods/mod.ts";
import type { StandardMethods } from "./standard-methods.ts";

/**
 * standardMethodsWithDenoKv is a decorator factory that creates a decorator
 * that adds Deno.Kv support to the standard Create, Delete, Get, List, and
 * Update operations of the resource for persistent storage.
 */
export function standardMethodsWithDenoKv(kv?: Deno.Kv): StandardMethods {
  return {
    create: (options) => standardCreate({ kv, ...options }),
    delete: (options) => standardDelete({ kv, ...options }),
    get: (options) => standardGet({ kv, ...options }),
    list: (options) => standardList({ kv, ...options }),
    update: (options) => standardUpdate({ kv, ...options }),
  };
}
