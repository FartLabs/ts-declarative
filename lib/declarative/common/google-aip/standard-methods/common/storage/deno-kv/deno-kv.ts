import type { StandardMethodStorage } from "#/lib/declarative/common/google-aip/standard-methods/common/storage/standard-method-storage.ts";

/**
 * DenoKvStandardMethodStorage is a storage system that uses Deno Kv to store data.
 */
export class DenoKvStandardMethodStorage implements StandardMethodStorage {
  /**
   * constructor constructs a new DenoKvStandardMethodStorage.
   */
  public constructor(public kv: Deno.Kv) {}

  /**
   * set sets the value of a key into Deno Kv.
   */
  public async set<T>(key: string[], value: T): Promise<void> {
    const result = await this.kv.set(key, value);
    if (!result.ok) {
      throw new Error(`Deno Kv: Failed to set key ${key}`);
    }
  }

  /**
   * get gets the value of a key from Deno Kv.
   */
  public async get<T>(key: string[]): Promise<T | null> {
    const result = await this.kv.get<T>(key);
    return result.value;
  }

  /**
   * delete deletes the value of a key from Deno Kv.
   */
  public async delete(key: string[]): Promise<void> {
    await this.kv.delete(key);
  }

  /**
   * list lists the values from Deno Kv.
   */
  public async *list<T>(prefix: string[] = []): AsyncIterable<T> {
    for await (const entry of this.kv.list<T>({ prefix })) {
      yield entry.value;
    }
  }
}
