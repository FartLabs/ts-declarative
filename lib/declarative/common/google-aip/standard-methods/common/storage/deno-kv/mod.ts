export * from "./deno-kv.ts";
