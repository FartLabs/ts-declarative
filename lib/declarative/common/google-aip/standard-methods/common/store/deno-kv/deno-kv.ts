import type { StandardMethodStore } from "#/lib/declarative/common/google-aip/standard-methods/common/store/standard-method-store.ts";

/**
 * DenoKvStandardMethodStore is a storage system that uses Deno Kv to store data.
 */
export class DenoKvStandardMethodStore implements StandardMethodStore {
  /**
   * constructor constructs a new DenoKvStandardMethodStore.
   */
  public constructor(public kv: Deno.Kv) {}

  /**
   * set sets the value of a key into Deno Kv.
   */
  public async set<T>(key: string[], value: T): Promise<void> {
    const result = await this.kv.set(key, value);
    if (!result.ok) {
      throw new Error(`Deno Kv: Failed to set key ${key}`);
    }
  }

  /**
   * get gets the value of a key from Deno Kv.
   */
  public async get<T>(key: string[]): Promise<T | null> {
    const result = await this.kv.get<T>(key);
    return result.value;
  }

  /**
   * delete deletes the value of a key from Deno Kv.
   */
  public async delete(key: string[]): Promise<void> {
    await this.kv.delete(key);
  }

  /**
   * list lists the values from Deno Kv.
   */
  public async *list<T>(prefix: string[] = []): AsyncIterable<T> {
    for await (const entry of this.kv.list<T>({ prefix })) {
      yield entry.value;
    }
  }
}
