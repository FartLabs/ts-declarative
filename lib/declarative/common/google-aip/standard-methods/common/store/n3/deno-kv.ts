import { getAsBlob, set as setAsBlob } from "@kitsonk/kv-toolbox/blob";
// deno-types="@types/n3"
import type { Quad } from "n3";
// deno-types="@types/n3"
import { Parser, Store, Writer } from "n3";

/**
 * getAsN3Store gets an n3 store from Deno.Kv storage.
 */
export async function getAsN3Store(
  kv: Deno.Kv,
  key: Deno.KvKey,
  consistency?: Deno.KvConsistencyLevel,
): Promise<Store<Quad, Quad, Quad, Quad> | null> {
  const store = new Store<Quad, Quad, Quad, Quad>();
  const file = await getAsBlob(kv, key, { consistency });
  if (file !== null) {
    // https://github.com/rdfjs/N3.js#parsing
    const parser = new Parser({ format: file.type || undefined });
    const quads = parser.parse(await file.text());
    store.addQuads(quads);
    return store;
  }

  return null;
}

/**
 * setAsN3Store sets an n3 store in Deno.Kv storage.
 */
export async function setAsN3Store(
  kv: Deno.Kv,
  key: Deno.KvKey,
  store: Store<Quad, Quad, Quad, Quad>,
  options?: { expireIn?: number; format?: string },
): Promise<Deno.KvCommitResult> {
  // https://github.com/rdfjs/N3.js#writing
  const writer = new Writer({ format: options?.format });
  const result = writer.quadsToString(store.toArray());
  const blob = new Blob([new TextEncoder().encode(result)], {
    type: options?.format,
  });

  return await setAsBlob(kv, key, blob, { expireIn: options?.expireIn });
}
