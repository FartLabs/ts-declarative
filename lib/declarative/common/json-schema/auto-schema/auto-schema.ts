import { Project } from "ts-morph";
import type { Class } from "#/lib/declarative/declarative.ts";
import { createDecoratorFactory } from "#/lib/declarative/decorator.ts";
import { declarativeTypeInfo } from "#/lib/declarative/common/type-info/type-info.ts";
import type { JSONSchemaMask } from "#/lib/declarative/common/json-schema/json-schema.ts";
import { declarativeJSONSchema } from "#/lib/declarative/common/json-schema/json-schema.ts";

/**
 * createAutoSchemaDecoratorFactory creates a decorator factory that decorates
 * a value with a JSON Schema and automatically reads the type information
 * from the class.
 */
export async function createAutoSchemaDecoratorFactoryAt(
  { url }: ImportMeta,
): Promise<(mask?: JSONSchemaMask) => (target: Class) => Class> {
  const project = new Project({ useInMemoryFileSystem: true });
  const specifier = new URL(url);
  project.createSourceFile(
    specifier.toString(),
    await Deno.readTextFile(specifier),
  );

  return createDecoratorFactory({
    initialize: (mask?: JSONSchemaMask) => {
      return [
        declarativeTypeInfo(project, specifier),
        declarativeJSONSchema(mask),
      ];
    },
  });
}
