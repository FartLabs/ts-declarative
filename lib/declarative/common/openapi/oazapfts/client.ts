import type { Opts } from "oazapfts";
import { generateSource } from "oazapfts";
import type { Class } from "#/lib/declarative/declarative.ts";
import { getPrototypeValue } from "#/lib/declarative/declarative.ts";
import type { ValueOpenAPISpecification } from "#/lib/declarative/common/openapi/specification.ts";

/**
 * generateOazapftsClientOf returns the source code generated by Oazapfts.
 *
 * @example
 * ```ts
 * @standardCreate()
 * @standardGet()
 * @autoSchema()
 * class Person {
 *   public constructor(public name: string) {}
 * }
 *
 * @openapi({ resources: [Person] })
 * class App {}
 *
 * const sourceCode = await generateOazapftsClientOf(App);
 * await Deno.writeTextFile("client.ts", sourceCode);
 * ```
 */
export async function generateOazapftsClientOf(
  target: Class,
  options?: OazapftsOptions,
): Promise<string | undefined> {
  return await generateSource(
    getPrototypeValue<ValueOpenAPISpecification>(target)
      ?.specification as unknown as string,
    options,
  );
}

/**
 * OazapftsOptions is the options for the Oazapfts decorator.
 */
export type OazapftsOptions = Opts;
