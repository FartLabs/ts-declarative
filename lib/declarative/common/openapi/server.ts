import type { Class, Declarative } from "#/lib/declarative/declarative.ts";
import { createDecoratorFactory } from "#/lib/declarative/decorator.ts";
import type {
  RouterDecoratorOptions,
  ValueRouterRoutes,
} from "#/lib/declarative/common/router/router.ts";
import { declarativeRoutes } from "#/lib/declarative/common/router/router.ts";
import type {
  OpenAPISpecificationDecoratorOptions,
  ValueOpenAPISpecification,
} from "./specification.ts";
import { declarativeOpenAPISpecification } from "./specification.ts";

/**
 * openapi is the decorator for OpenAPI server.
 */
export const openapi: (
  options?: OpenAPIServerDecoratorOptions | undefined,
) => (target: Class) => Class = createOpenAPIServerDecoratorFactory();

/**
 * createOpenAPIServerDecoratorFactory is the factory function for the
 * OpenAPI server decorator.
 */
export function createOpenAPIServerDecoratorFactory(): (
  options?: OpenAPIServerDecoratorOptions | undefined,
) => (target: Class) => Class {
  return createDecoratorFactory({
    initialize: (
      options?: OpenAPIServerDecoratorOptions,
    ): Array<Declarative<ValueOpenAPIServer>> => {
      return [
        declarativeOpenAPISpecification(options),
        declarativeRoutes(options),
      ];
    },
  });
}

/**
 * OpenAPIServerDecoratorOptions is the options for the OpenAPI server
 * decorator.
 */
export interface OpenAPIServerDecoratorOptions
  extends OpenAPISpecificationDecoratorOptions, RouterDecoratorOptions {}

/**
 * ValueOpenAPIServer is the value associated with an OpenAPI server.
 */
export interface ValueOpenAPIServer
  extends ValueOpenAPISpecification, ValueRouterRoutes {}
