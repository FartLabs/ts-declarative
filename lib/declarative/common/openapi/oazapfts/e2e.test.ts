import { assertEquals } from "@std/assert/equals";
import { assertSnapshot } from "@std/testing/snapshot";
import { openapi } from "#/lib/declarative/common/openapi/server.ts";
import { standardMethodsWithDenoKv } from "#/lib/declarative/common/google-aip/deno-kv.ts";
import { createAutoSchemaDecoratorFactoryAt } from "#/lib/declarative/common/json-schema/auto-schema/auto-schema.ts";
import { generateOazapftsClientOf } from "./client.ts";
import { routerOf } from "#/lib/declarative/common/router/router.ts";

const autoSchema = await createAutoSchemaDecoratorFactoryAt(import.meta);

const kv = await Deno.openKv(":memory:");
const standardMethod = standardMethodsWithDenoKv(kv);

@standardMethod.create()
@standardMethod.get()
@autoSchema()
class Person {
  public constructor(public name: string) {}
}

@openapi({
  specification: { servers: [{ url: "http://localhost:8080" }] },
  resources: [Person],
})
class App {}

Deno.test(
  "generateOazapftsClientOf generates client from OpenAPI specification",
  async (t) => {
    const sourceCode = await generateOazapftsClientOf(
      App,
      { optimistic: true },
    );
    await assertSnapshot(t, sourceCode);
  },
);

Deno.test("e2e Oazapfts client dynamic import", async () => {
  const server = Deno.serve({ port: 8080 }, routerOf(App));
  const ash = new Person("Ash Ketchum");

  const sourceCode = await generateOazapftsClientOf(App, { optimistic: true });
  if (sourceCode === undefined) {
    throw new Error("sourceCode is undefined");
  }

  const client = await import(
    `data:text/typescript;base64,${btoa(sourceCode)}`
  );

  const createdPerson = await client.postPeople(ash);
  assertEquals(createdPerson.name, ash.name);

  const fetchedPerson = await client.getPeopleByPerson(ash.name);
  assertEquals(fetchedPerson.name, ash.name);

  await server.shutdown();
});
