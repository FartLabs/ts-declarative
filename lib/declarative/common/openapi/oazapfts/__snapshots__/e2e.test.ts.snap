export const snapshot = {};

snapshot[`generateOazapftsClientOf generates client from OpenAPI specification 1`] = `
'/**
 * App
 * 1.0.0
 * DO NOT MODIFY - This file has been generated using oazapfts.
 * See https://www.npmjs.com/package/oazapfts
 */
import * as Oazapfts from "@oazapfts/runtime";
import * as QS from "@oazapfts/runtime/query";
export const defaults: Oazapfts.Defaults<Oazapfts.CustomHeaders> = {
    headers: {},
    baseUrl: "http://localhost:8080",
};
const oazapfts = Oazapfts.runtime(defaults);
export const servers = {
    server1: "http://localhost:8080"
};
export type Person = {
    name: string;
};
/**
 * Gets Person
 */
export function getPeopleByPerson(person: string, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: Person;
    }>(\`/people/\${encodeURIComponent(person)}\`, {
        ...opts
    }));
}
/**
 * Creates Person
 */
export function postPeople(person: Person, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: Person;
    }>("/people", oazapfts.json({
        ...opts,
        method: "POST",
        body: person
    })));
}
'
`;
